packages/flutter/lib/src/services/hardware_keyboard.dart

```diff
@@ -0,0 +1,894 @@
+  @visibleForTesting
+  void clearState() {
+    assert(() {
+      _pressedKeys.clear();
+      _lockModes.clear();
+      _handlers.clear();
+      assert(_modifiedHandlers == null);
+      return true;
+    }());
+  }
+}
+
+/// The mode in which information of key events is delivered.
+///
+/// Different platforms deliver key events in different modes, which differ
+/// in dispatching functions, event classes, and event models. These modes are
+/// called "key event vehicles".
+///
+/// Determining the platform vehicle allows both sets of framework API (those
+/// related to [RawKeyEvent] and those to [KeyEvent]) to work on all platforms
+/// regardless of their implementation.  The vehicle of the current platform is
+/// inferred at the start of the application, and is used to guide how key
+/// event information is interpreted and dispatched, in order for the
+/// higher-level APIs to behave in the same way regardless of platform
+/// vehicles.
+///
+/// See also:
+///
+///  * [KeyEventManager], which infers platform vehicle and guides
+///    how key messages are dispatched.
+///  * [KeyEventSimulator.vehicle], the vehicle used to simulate key events
+///    in tests, controlled by [debugKeySimulationVehicleOverride].
+///  * [debugKeySimulationVehicleOverride], the variable to control
+///    [KeyEventSimulator.vehicle].
+///  * [KeySimulationVehicleVariant], an easier way to set
+///    [debugKeySimulationVehicleOverride] in widget tests.
+enum KeyEventVehicle {
+  /// Key event information is delivered as raw key data.
+  ///
+  /// Raw key data is platform's native key event information sent in JSON
+  /// through a method channel, which is then interpreted as a platform subclass
+  /// of [RawKeyEventData].
+  ///
+  /// If the current vehicle is [rawKeyData], the raw key data is converted to
+  /// both [KeyMessage.events] and [KeyMessage.rawEvent].
+  rawKeyData,
+
+  /// Key event information is delivered as converted key data, followed
+  /// by raw key data.
+  ///
+  /// Key data ([ui.KeyData]) is a standardized event stream converted from
+  /// platform's native key event information, sent through the embedder
+  /// API. Its event model is described in [HardwareKeyboard].
+  ///
+  /// Raw key data is platform's native key event information sent in JSON
+  /// through a method channel. It is interpreted by subclasses of
+  /// [RawKeyEventData].
+  ///
+  /// If the current vehicle is [rawKeyData], the key data is converted to
+  /// [KeyMessage.events], and the raw key data is converted to
+  /// [KeyMessage.rawEvent].
+  keyDataThenRawKeyData,
+}
+
+/// The assumbled information corresponding to a native key message.
+///
+/// While Flutter's [KeyEvent]s are converted from key messages from the native
+/// platform, every native message might result in multiple [KeyEvent]s, for
+/// example, in order to synthesize missed modifier key presses or releases.
```

Comment:
```suggestion
/// platform, every native message might result in multiple [KeyEvent]s. For
/// example, this might happen in order to synthesize missed modifier key
/// presses or releases.
```